backends/bmv2/common/header.cpp

```diff
@@ -135,18 +135,41 @@ void HeaderConverter::addHeaderStacks(const IR::Type_Struct* headersStruct) {
         auto stack_name = f->controlPlaneName();
         auto stack_size = stack->getSize();
         auto type = ctxt->typeMap->getTypeType(stack->elementType, true);
-        BUG_CHECK(type->is<IR::Type_Header>(), "%1% not a header type", stack->elementType);
-        auto ht = type->to<IR::Type_Header>();
-        addHeaderType(ht);
-        cstring stack_type = stack->elementType->to<IR::Type_Header>()
-                                               ->controlPlaneName();
-        std::vector<unsigned> ids;
-        for (unsigned i = 0; i < stack_size; i++) {
-            cstring hdrName = f->controlPlaneName() + "[" + Util::toString(i) + "]";
-            auto id = ctxt->json->add_header(stack_type, hdrName);
-            ids.push_back(id);
+        bool isUnion = type->is<IR::Type_HeaderUnion>();
+        if (type->is<IR::Type_Header>() || isUnion) {
+            auto ht = type->to<IR::Type_StructLike>();
+            addHeaderType(ht);
+            cstring stack_type = ht->controlPlaneName();
+            std::vector<unsigned> ids;
+            for (unsigned i = 0; i < stack_size; i++) {
+                cstring hdrName = f->controlPlaneName() + "[" + Util::toString(i) + "]";
+                int id;
+                if (isUnion) {
+                    // We have to add separately a header instance for all
+                    // headers in the union.  Each instance will be named with
+                    // a prefix including the union name, e.g., "u.h"
+                    Util::JsonArray* fields = new Util::JsonArray();
+                    for (auto uf : type->to<IR::Type_HeaderUnion>()->fields) {
```

Comment:
maybe we could have a helper function for that for loop, since it seems you have the exact same loop below (for `HeaderConverter::init_apply`)